accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,341 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+    inferred type** on all other reachable paths is `Null`, then the
+    **initialization inferred type** after the join is `T?`.
+  - Otherwise, it is an error.
+
+In other words, if an **untyped** variable of any kind is assigned on all paths
+into the join, it must be assigned either a value of the same type, or a value
+of type `Null`, on all paths; and if the variable is assigned a value of type
+`Null` on any path, then the variable is treated as having nullable type.
+
+At a join point in the program, if an **untyped variable** has been given an
+**initialization inferred type** on some but not all of the reachable paths into
+the join, then the **initialization inferred type** for the variable after the
+join point is computed as follows:
+  - If the variable is declared `final` and not `late` it is an error, and there
+    is no **initialization inferred type** after the join.
+  - If the variable is declared `late` and the **initialization inferred type**
+    on each path which has one is the same type `T` (using syntactic equality),
+    then the **initialization inferred type** after the join is `T`.
+  - If the variable is declared `late` and/or `var` and the **initialization
+    inferred type** on one or more paths is the same type `T` (using syntactic
+    equality), and the **initialization inferred type** on one or more paths is
+    `Null`, then the **initialization inferred type** after the join is `T?`.
+  - If the variable is declared `var` and not `late` and the **initialization
+    inferred type** on one or more paths is the same type `T` (using syntactic
+    equality), and the **initialization inferred type** on all other paths (if
+    any) is `Null`, then the **initialization inferred type** after the join is
+    `T?`.
+  - Otherwise it is an error.
+
+In other words, a non-`late`, `final` **untyped** variable must be assigned a
+value on every path, but a variable declared `late` and/or `var` may be assigned
+a value on only some of the paths into a join.  In the latter case, the type is
+made nullable if the variable is declared as non-`late`, and the variable is
+treated as implicitly null initialized.  Variables declared `late` are only made
+nullable if `Null` is explicitly assigned on one more paths into the join, and
+there is no implicit null initialization.  In all cases, the values assigned
+must all have the same type, or type `Null`.
+
+
+```dart
+void test(bool b) {
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = 3.0;
+    } // Error: x has two different types on different paths
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = null;
+    } // x has type int?
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int?
```

Comment:
I think this will *not* be what users want most of the time. It's probably a mistake that they didn't initialize it. But the idea is that you'll get a nullable type and then you'll get some other downstream error when you try to use that, so erroneous uses of this should still be caught. And then it *is* what you want, it lets you. Is that the idea?

If so, SGTM.

Reply:
Yes.  I mean, it's easy to not allow this at all - I can say that you *must* write a value on every path to the join, so you'd have to write this with an `x=null` in an else branch.  But the idea was to allow this, since anything else feels a little verbose.  The one place I can see this not catching legit errors is if you're using a nullable API, but you didn't actually want to pass `null`.

Reply:
I keep thinking about this case and I still worry that we're implicitly giving a user a nullable type that they didn't request. I wish we could instead make this case an error and tell them to solve it by writing:

```dart
var x = null; // User opts in to null.
if (b) {
  x = 3;
}
```

This is just a hunch, though, and I could be wrong. Here's the evidence I have in mind:

* Users often tell us that they want more "missing return" errors even in functions that can return `null`. Many seem to want this to be an error:

    ```dart
    Object foo(bool c) { // Object? with NNBD.
      if (c) return "thing";
    }
    ```
    
    And instead want to have to write:

    ```dart
    Object foo(bool c) { // Object? with NNBD.
      if (c) return "thing";
      return null;
    }
    ```

* Dart requires final fields to be definitely initialized in the constructor initialization list, even when `null` is a valid value. Users don't seem to mind having to put an explicit `= null` there. This may be confounded by them having different preferences between final and non-final fields, but I'm not sure about that.

* I've definitely had a number of users over the years tell me they are surprised by the Effective Dart rule that tells them to rely on default initialization to `null`. They felt it should have been made explicit.